src/inmanta/ast/type.py

```diff
@@ -612,11 +821,34 @@ def validate(self, value: object) -> bool:
     def type_string_internal(self) -> str:
         return "Union[%s]" % ",".join(t.type_string_internal() for t in self.types)
 
+    def as_python_type_string(self) -> "str | None":
+        types = [tp.as_python_type_string() for tp in self.types]
+
+        effective_types = [tp for tp in types if tp is not None]
+        if len(types) != len(effective_types):
+            # One is not converted
+            return None
+        return " | ".join(effective_types)
+
+    def has_custom_to_python(self) -> bool:
+        # If we mix convertible and non convertible, it won't work, so we avoid it
+        return False
+
+    def corresponds_to(self, type: Type) -> bool:
+        raise NotImplementedError("No unions can be specified as attributes, unexpected usage")
+
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, Union):
             return NotImplemented
         return self.types == other.types
 
+    def is_primitive(self) -> bool:
+        return all(tp.is_primitive() for tp in self.types)
```

Comment:
This makes types like `int | list[int | str] | dict[str, float | str]` (python syntax) primitive as well. Is that intentional?

Reply:
yes, under this definition of primitive.

Reply:
But we don't allow them for dataclasses, purely because they can't be expressed in the DSL?